XenAdmin/ConsoleView/XSVNCScreen.cs

```diff
@@ -474,100 +474,118 @@ private void SetPendingVNCConnection(Stream s)
             }
         }
 
-        private bool scaling;
+        private bool _scaling;
+
         public bool Scaling
         {
             get
             {
                 Program.AssertOnEventThread();
-                return scaling;
+                return _scaling;
             }
 
             set
             {
                 Program.AssertOnEventThread();
-                scaling = value;
+                _scaling = value;
                 if (RemoteConsole != null)
                     RemoteConsole.Scaling = value;
             }
         }
 
         public IRemoteConsole RemoteConsole
         {
-            get { return vncClient != null ? (IRemoteConsole)vncClient : rdpClient; }
-            set 
+            get => _vncClient != null ? (IRemoteConsole)_vncClient : _rdpClient;
+            set
             {
-                if (vncClient != null) 
-                    vncClient = (VNCGraphicsClient) value ;
-                else if (rdpClient != null) 
-                    rdpClient = (RdpClient)value;
+                if (_vncClient != null)
+                    _vncClient = (VNCGraphicsClient)value;
+                else if (_rdpClient != null)
+                    _rdpClient = (RdpClient)value;
             }
         }
 
+        private readonly object _rdpConnectionLock = new object();
+
+
         /// <summary>
         /// Creates the actual VNC or RDP client control.
         /// </summary>
-        private void initSubControl()
+        private void InitSubControl()
         {
             Program.AssertOnEventThread();
 
             //When switch to RDP from VNC, if RDP IP is empty, do not try to switch.
-            if (String.IsNullOrEmpty(RdpIp) && !UseVNC && RemoteConsole != null)
+            if (string.IsNullOrEmpty(RdpIp) && !UseVNC && RemoteConsole != null)
                 return;
 
-            bool wasFocused = false;
-            this.Controls.Clear();
+            var wasFocused = false;
+            Controls.Clear();
             //console size with some offset to accomodate focus rectangle
-            Size currentConsoleSize = new Size(this.Size.Width - CONSOLE_SIZE_OFFSET, this.Size.Height - CONSOLE_SIZE_OFFSET);
-                
-            // Stop the old client.
-            if (RemoteConsole != null)
+            var currentConsoleSize = new Size(Size.Width - CONSOLE_SIZE_OFFSET, Size.Height - CONSOLE_SIZE_OFFSET);
+
+            lock (_rdpConnectionLock)

```

Comment:
Why is the lock needed? The method is always called on the main thread (there is an assertion at the beginning of the method), so it's not likely that two threads will be accessing it.

Reply:
I thought it might prevent future issues. In general my approach was to try and limit multi-thread access to `RemoteConsole` so this is just part of that.

Reply:
> If a different thread accesses this code, the assertion will throw a Fatal error before you get to accessing the RemoteConsole.
good point 👍 will remove it